doc/code_coverage_rate/code_coverage_rate.md

```diff
@@ -0,0 +1,173 @@
+#  [Code Coverage Rate HLD] Update code coverage rate.md
+code-coverage-rate-hld
+
+# About this Document
+This document provides a high level of design for code coverage rate testing.
+
+# Problem Definition
+Code coverage rate is a term that describes the rate of the number of code lines that have been executed divide by the number of total code lines in the switch operating system. We reuse this term in SONiC code testing and create some related tools for SONiC code testing.
+
+# Solution
+Use the Gcov tools to measure the code coverage rate in its commercial switch and white box switch. Create the Linux scripts for Gcov compiling, Gcov data collect, and Gcov data show functions for SONiC. A GCOV\_ENABLE compiling tag is provided for the SONiC code project. And we integrate all these Gcov functions into the Jenkins CICD system. Our solution is a complete toolset for the code coverage rate for SONiC.
```

Comment:
is there a cpu performance overhead when gcov is enabled? If yes, what is the additional cpu perforamnce overhead?

Reply:
Currently we didn't find any cpu usage overhead when compiling or running a gcov-versioned image. We will provide a data comparison in the next revision

Reply:
Hi lguohan, below is the comparison for cpu usage overhead, please check:
'''

Unker 4k ARP flow: 
root@switch2:/home/admin# arp -n | wc -l
4002

cpu usage overhead for normal image, the average is about 13.95:
root@switch2:/home/admin# sar -u 1 100
03:02:23 AM     CPU     %user     %nice   %system   %iowait    %steal     %idle
03:02:24 AM     all     11.55      0.00     21.83      0.00      0.00     66.62
03:02:25 AM     all      5.25      0.00     20.61      0.00      0.00     74.14
03:02:26 AM     all      6.52      0.00     24.84      0.00      0.00     68.63
......
03:04:02 AM     all      3.23      0.00     13.55      0.00      0.00     83.23
03:04:03 AM     all      3.06      0.00     12.76      0.00      0.00     84.18
Average:        all      3.65      0.00     13.95      0.02      0.00     82.39

cpu usage overhead for GCOV image, the average is about 15.55:
root@switch2:/home/admin# sar -u 1 100
12:58:36 AM     CPU     %user     %nice   %system   %iowait    %steal     %idle
12:58:37 AM     all      4.64      0.00     14.55      0.00      0.00     80.80
12:58:38 AM     all      2.16      0.00     14.97      0.13      0.00     82.74
......
01:00:14 AM     all      2.29      0.00     13.76      0.00      0.00     83.95
01:00:15 AM     all      8.48      0.00     18.70      0.00      0.00     72.82
01:00:16 AM     all      1.65      0.00     12.80      0.00      0.00     85.55
Average:        all      5.75      0.00     15.55      0.02      0.00     78.68

'''